soufi/finders/photon.py

```diff
@@ -0,0 +1,160 @@
+        # have one more way to look this up using the 'repomd' package
+        ref = self._repo_lookup(f"{PHOTON_PACKAGES}{dir_}")
+        if url and ref:
+            return PhotonDiscoveredSource([f'{url}{ref}'])
+        raise exceptions.SourceNotFound
+
+    def _get_url(self, url):
+        response = requests.get(url, timeout=TIMEOUT)
+        if response.status_code != requests.codes.ok:
+            raise exceptions.DownloadError(response.reason)
+        return response.content
+
+    def _get_dir(self, url=PHOTON_PACKAGES):
+        """Get the relevant package dir."""
+        content = self._get_url(url)
+        tree = html.fromstring(content)
+        dirs = tree.xpath('//a/text()')
+        dirs = [d for d in dirs if d[0].isdigit()]
+
+        # Do very basic checks to see if we can limit searched dirs based on
+        # the package version containing a 'phN' hint.
+        el = re.search(r'.ph(\d)', self.version)
+        if el is not None:
+            dirs = [d for d in dirs if d[0] == el.expand(r'\1')]
+
+        # We should only have 1 candidate directory left.  If not, give up,
+        # because something's gone wrong and we're not going to find it
+        if len(dirs) != 1:
+            raise exceptions.SourceNotFound
+        return dirs[0]
+
+    def _get_source_dirs_content(self, release_dir):
+        """Yield all content from dirs under release_dir that have source.
+
+        A Release dir is the dir under the top-level PHOTON_PACKAGES.
+        Under there, there are many directories that can contain packages,
+        and this function yields each in turn.
+        """
+        url = f"{PHOTON_PACKAGES}{release_dir}"
+        content = self._get_url(url)
+        tree = html.fromstring(content)
+        # Ideally all the SRPM trees would have the exact same packages in
+        # them, but their `aarch64` trees seem to be a little light.  Prefer
+        # x86_64 to be safe
+        dirs = tree.xpath(
+            "//a[text()[contains(.,'srpms')][contains(.,'x86_64')]]/text()"
+        )
+
+        for dir_ in dirs:
+            try:
+                _url = f"{url}{dir_}"
+                yield _url, self._get_url(_url)
+            except exceptions.DownloadError:
+                pass
+
+    def _get_path(self, content):
+        """Given a dir, see if it contains the source HREF."""
+        # Grab source directory listing at dir and look for
+        # <name>-<version>.src.rpm.
+        tree = html.fromstring(content)
+        href = f"{self.name}-{self.version}.src.rpm"
+        ref = tree.xpath('//a[@href=$href]', href=href)
+        if ref:
+            return href
+        return None
+
+    def _repo_lookup(self, base_url):
+        content = self._get_url(base_url)
+        tree = html.fromstring(content)
+        # micro-optimization: the dirs come back in essentially the reverse
+        # order of likelihood of containing the package.  Walk them in
+        # reverse order.
```

Comment:
It won't break anything but this seems like a flaky approach to determining an order as it depends on the remote page. It might be better to be explicit about which dirs you want to process in what order.

Reply:
The pages are auto-generated (and poorly, most of the index pages simply stop after a certain size) and alpha-sorted, so it was a safe assumption.

But these things will be cached, and they're small, so there's no harm in just walking the list in the returned order.  I'll drop it.